source/sigv4.c

```diff
@@ -794,6 +918,308 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+                                              uint32_t flags,
+                                              size_t * headerCount,
+                                              CanonicalContext_t * canonicalRequest )
+    {
+        size_t index = 0, noOfHeaders;
+        const char * pKeyOrValStartLoc;
+        const char * pCurrLoc;
+        bool keyFlag = true;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( pHeaders != NULL );
+        assert( headersDataLen > 0 );
+        assert( canonicalRequest != NULL );
+        assert( headerCount != NULL );
+
+        noOfHeaders = *headerCount;
+        pKeyOrValStartLoc = pHeaders;
+        pCurrLoc = pHeaders;
+
+        for( index = 0; index < headersDataLen; index++ )
+        {
+            if( noOfHeaders == SIGV4_MAX_HTTP_HEADER_COUNT )
+            {
+                sigV4Status = SigV4MaxHeaderPairCountExceeded;
+                break;
+            }
+            /* Look for key part of an header field entry. */
+            else if( ( keyFlag ) && ( pHeaders[ index ] == ':' ) )
+            {
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData = pKeyOrValStartLoc;
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].key.dataLen = ( pCurrLoc - pKeyOrValStartLoc );
+                pKeyOrValStartLoc = pCurrLoc + 1U;
+                keyFlag = false;
+            }
+            /* Look for header value part of a header field entry for both canonicalized and non-canonicalized forms. */
+            /* Non-canonicalized headers will have header values ending with "\r\n". */
+            else if( ( !keyFlag ) && !( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) && ( ( index + 1 ) < headersDataLen ) &&
+                     ( 0 == strncmp( pCurrLoc, "\r\n", strlen( "\r\n" ) ) ) )
+            {
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = pKeyOrValStartLoc;
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( pCurrLoc - pKeyOrValStartLoc );
+                /* Set starting location of the next header key string after the "\r\n". */
+                pKeyOrValStartLoc = pCurrLoc + 2U;
+                keyFlag = true;
+                noOfHeaders++;
+            }
+            /* Canonicalized headers will have header values ending just with "\n". */
+            else if( ( !keyFlag ) && ( ( flags & SIGV4_HTTP_HEADERS_ARE_CANONICAL_FLAG ) && ( pHeaders[ index ] == '\n' ) ) )
+            {
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = pKeyOrValStartLoc;
+                canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( pCurrLoc - pKeyOrValStartLoc );
+                /* Set starting location of the next header key string after the "\n". */
+                pKeyOrValStartLoc = pCurrLoc + 1U;
+                keyFlag = true;
+                noOfHeaders++;
+            }
+
+            pCurrLoc++;
+        }
+
```

Comment:
It would be important to verify that each header key has found its header value by checking that the `keyFlag` value is reset to `true` (unless it is fine to have headers field without any value).